File: leetcode/src/main/golang/three_sum_closest.go
import "sort"

func threeSumClosest(nums []int, target int) int {
    if len(nums) < 3 {
        return 0
    }
    sort.Ints(nums)
    sum, curShift, minShift := 0, 0, 0x7fffffff
    for i := 0; i < len(nums) - 2; i++ {
        j, k, tar := i + 1, len(nums) - 1, target - nums[i]
        for ; j < k; {
            curShift = nums[j] + nums[k] - tar
            abs := curShift
            if abs < 0 {
                abs = -abs
            }
            if abs < minShift {
                minShift, sum = abs, nums[i] + nums[j] + nums[k]
            }
            if curShift == 0 {
                return target
            } else if curShift < 0 {
                j++
            } else {
                k--
            }
        }
    }
    return sum
}

File: leetcode/src/main/ruby/three_sum_closest.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
def three_sum_closest(nums, target)
    return 0 if nums.length < 3
    nums.sort!
    sum, c_shift, m_shift = 0, 0, 0x7fffffff
    for i in 0...nums.length - 2
        j, k, tar = i + 1, nums.length - 1, target - nums[i]
        while j < k
            c_shift = nums[j] + nums[k] - tar
            abs = c_shift < 0 ? -c_shift : c_shift
            sum, m_shift = nums[i] + nums[j] + nums[k], abs if abs < m_shift
            if c_shift == 0
                return target
            elsif c_shift < 0
                j += 1
            else
                k -= 1
            end
        end
    end
    sum
end